polaris/model_step.py

```diff
@@ -380,47 +333,41 @@ def add_streams_file(self, package, streams, template_replacements=None):
             dict(package=package, streams=streams,
                  replacements=template_replacements))
 
-    def update_streams_at_runtime(self, package, streams,
-                                  template_replacements):
+    def dynamic_model_config(self, at_setup):
         """
-        Update the streams files during the run phase of this step using the
-        given template and replacements.  This may be useful for updating
-        streams based on config options that a user may have changed before
-        running the step.
+        Add model config options, namelist, streams and yaml files using config
+        options or template replacements that need to be set both during step
+        setup and at runtime
 
         Parameters
         ----------
-        package : Package
-            The package name or module object that contains the streams file
-
-        streams : str
-            The name of a Jinja2 template to be rendered with replacements
-
-        template_replacements : dict
-            A dictionary of replacements
+        at_setup : bool
+            Whether this method is being run during setup of the step, as
+            opposed to at runtime
         """
-        if template_replacements is not None:
-            print(f'Warning: updating streams in {self.streams} using the '
-                  f'following template and replacements:')
-            print(f'{package} {streams}')
-            for key, value in template_replacements.items():
-                print(f'{key} = {value}')
-
-        filename = os.path.join(self.work_dir, self.streams)
-
-        tree = etree.parse(filename)
-        tree = polaris.streams.read(package, streams, tree=tree,
-                                    replacements=template_replacements)
-        polaris.streams.write(tree, filename)
+        if self.update_pio and not at_setup:
+            self.update_namelist_pio()
```

Comment:
I could see how this might have made sense but in this case I don't think we ant to print anything.  This is a case where calling `self.update_namelist_pio()` only makes sense once the resources that will actually be used are established.  I tried calling it in both cases and got values based on the target `ntasks` and that felt quite confusing to me.

We have not tried to run `update_namelist_pio()` during setup in compass so I think this will be an intuitive behavior in polaris, too.

Also, I am very reluctant to print anything at all during setup because it goes directly to the terminal and complicates what is currently a very simple, clean list of test cases.